### PR TITLE
Bluestore/NVMEDevice: Update SPDK to 17.10

### PR DESCRIPTION
**Pull request by optimistyzy**
Need to add uuid lib in cmake file due to upgrade.
Test this patch in a machine with fio benchmark and it
works well.

Signed-off-by: Ziye Yang <optimistyzy@gmail.com>

### CODE REVIEW

**Comment by optimistyzy** [comment]
@yuyuyu101

**Comment by yuyuyu101** [comment]
it's ok now

**Comment by tchaikov** [comment]
retest this please.

File: cmake/modules/BuildSPDK.cmake [hunk 1]
```diff
@@ -24,7 +24,7 @@ macro(build_spdk)
       INTERFACE_INCLUDE_DIRECTORIES "${source_dir}/include")
     list(APPEND SPDK_LIBRARIES spdk::${c})
   endforeach()
-  list(APPEND iface_libs ${DPDK_LIBRARIES} rt)
+  list(APPEND iface_libs ${DPDK_LIBRARIES} uuid rt)
```
Comment (tchaikov) [inline cmake/modules/BuildSPDK.cmake]: please note, uuid is not a dependency of `spdk::env_dpdk`, it is a dependency of `libspdk_lvol`. we need to 

1. detect libuuid-dev or using `find_package(libuuid REQUIRED)` unless we have a way to disable it, but AFAICT, it's not optional in SPDK. please refer to 3170f16e864e257416bccdd73777bff775767a8f
1. add libuuid-dev{,el} back to `ceph.spec.in` and `debian/control`.
1. add `lvol` to the component list at https://github.com/ceph/ceph/pull/19096/files#diff-05c598c78a24f8547e136669dd4429ccR19.
1. and add `INTERFACE_LINK_LIBRARIES` for `spdk::lvol`

Comment (tchaikov) [reply cmake/modules/BuildSPDK.cmake]: and the new DPDK version needs libnuma-dev. we need ensure it's either disabled or our building script can fulfill the requirement.

Comment (wjwithagen) [reply cmake/modules/BuildSPDK.cmake]: @optimistyzy @tchaikov 
Dependend on the calls it doew not require uuid to be included because routines are in libc.
Or it needs to be added as dependancy in make-deps.sh as well.
But then than would need to be the e2fsprogs-libuuid package

**Comment by optimistyzy** [comment]
hi @tchaikov， I am not familiar with ceph dependency. Could you help it? for spdk lvol dependency,  it is not optional.

**Comment by tchaikov** [comment]
@optimistyzy please file a ticket at tracker.

File: cmake/modules/BuildSPDK.cmake (hunk 1 above)
Comment (tchaikov) [reply cmake/modules/BuildSPDK.cmake]: @wjwithagen i don't know what "calls" or "routines" or "libc" you were looking at. but take `uuid_compare()` as an example, it is not part of POSIX, C standard,or any extension. see https://linux.die.net/man/3/uuid_compare.

> But then than would need to be the e2fsprogs-libuuid package

please feel free to do it. but please note that SPDK is disabled on FreeBSD.

Comment (tchaikov) [reply cmake/modules/BuildSPDK.cmake]: > The `libuuid` was removed from CMake files during one of the library sweep cleanups you did.

actually i was referencing that very commit in my reply of https://github.com/ceph/ceph/pull/19096#discussion_r152490524

> This is what the FreeBSD manpage gives for `man uuid`:

TIL !

Comment (optimistyzy) [reply cmake/modules/BuildSPDK.cmake]: @wjwithagen SPDK also supports FreeBSD

Comment (wjwithagen) [reply cmake/modules/BuildSPDK.cmake]: @optimistyzy 
I that a question or a remark? 
Yes, SPDK is supported on FreeBSD.
But the way you require libuuid will break on FreeBSD, because the `uuid_*` routines are already resolved by `libc`. So please only add `uuid` only for other platforms.

**Comment by optimistyzy** [comment]
@wjwithagen @tchaikov I changed to only link uuid in Linux platform, as @wjwithagen  uuid is already included in libc, so we do not need to link that.  Plus @tchaikov 's follwoing patch:
https://github.com/ceph/ceph/commit/3170f16e864e257416bccdd73777bff775767a8f
We should not break anything on FreeBSD.

**Comment by tchaikov** [comment]
> I changed to only link uuid in Linux platform

your latest change still links against libuuid on all platforms.

> Plus @tchaikov 's follwoing patch:
3170f16e864e257416bccdd73777bff775767a8f

i don't follow you. what are you referencing it in this context? let me rephrase: my patch is missing in your change. and i believe it's necessary to `find_package(foo)` before using `foo`. hence, your change is broken. and i elaborated the steps to fix it. please let me know if you need more info from my side.

**Comment by optimistyzy** [comment]
@tchaikov With your uuid find patch plus mine, will it still fail the automation test?

**Comment by tchaikov** [comment]
@optimistyzy probably you could test it by yourself locally? would be easier this way, IMO.

**Comment by optimistyzy** [comment]
local test already passed。。。。

File: CMakeLists.txt [hunk 2]
```diff
@@ -224,9 +224,15 @@ if(WITH_SPDK)
   if(NOT WITH_BLUESTORE)
     message(SEND_ERROR "Please enable WITH_BLUESTORE for using SPDK")
   endif()
-  include(BuildSPDK)
-  build_spdk()
-  set(HAVE_SPDK TRUE)
+  find_package(uuid REQUIRED)
```
Comment (tchaikov) [inline CMakeLists.txt]: `Finduuid.cmake` is not included by cmake or ceph. how come this passes your local test?

File: CMakeLists.txt [hunk 3]
```diff
@@ -224,9 +224,15 @@ if(WITH_SPDK)
...
+  set(HAVE_UUID ${UUID_FOUND})
+  if(HAVE_UUID TRUE)
+    include(BuildSPDK)
+    build_spdk()
+    set(HAVE_SPDK TRUE)
+  else
+   set(HAVE_SPDK FALSE)
```
Comment (tchaikov) [inline CMakeLists.txt]: please do not set `HAVE_SPDK` to FALSE when `WITH_SPDK` is ONE. if user wants to build a ceph `WITH_SPDK`, we should do so or just fail.

File: CMakeLists.txt (hunk 2 above)
Comment (optimistyzy) [reply CMakeLists.txt]: patch your pr， then patch mine

**Comment by tchaikov** [comment]
this PR is half backed. closing in favor of #19208

**Comment by optimistyzy** [comment]
@tchaikov great， thanks for fixing all spdk and dpdk related issues due to upgrade
